### ISSUE TITLE
future_map with boot function

### ISSUE THREAD

**Issue by Dokmen** [DavisVaughan/furrr#190]
Hi,

I'm studying microarray analysis with ~23000 genes with 31 samples. My aim is to finde expressed genes. I'm using boot funciton from boot package. The code is
`boot(data,  statistic = bootFun, R = 999, parallel = "multicore", ncpus =48)` 

With this  parallel = "multicore" option, the process is completed in 10 minutes in the 23000*31 matrix. When I converted matrix to vectorized nested format by group, I could not find good options for parallel process for boot. 

the code is vectorized format is 
`library(furrrr)
plan(multicore, workers=48)
df[, booted := future_map(.x=data, ~ boot(data=.x, sim = "ordinary", 
                                          statistic = bootFun, R = 999))]`

this code has been running for about an hour. It is very slow. I can't understand where I went wrong. Dou you have a suggestion?

**Comment by DavisVaughan** [comment]
Could you please turn this into a self-contained [reprex](http://reprex.tidyverse.org/) (short for minimal **repr**oducible **ex**ample)? It will help us help you if we can be sure we're all working with/looking at the same stuff.

If you've never heard of a reprex before, you might want to start by reading the tidyverse.org [help page](https://www.tidyverse.org/help/).

You can install reprex by running (you may already have it, though, if you have the tidyverse package installed):
```r
install.packages("reprex")
```

Thanks

**Comment by Akyanus** [comment]
``` r
library(dplyr)
library(tidyr)
library(stringr)
library(purrr)
library(data.table)
library(boot)
library(furrr)
#> Loading required package: future

#Expression matrix
mat <- data.frame(matrix(runif(600000),ncol=30))
colnames(mat) <- c(paste0("C", seq(1:10)), paste0("T", seq(11:30)))
rownames(mat) <- c(paste0("Gene", seq(1:20000)))

#function for boot
boot.fun <- function(data, i){
  data <- data[i,]
  t.obs <- apply(data, 1, function(x) tryCatch(t.test(x[11:30], x[1:10], alternative = "two.sided",
                                                      paired = FALSE, var.equal = FALSE)$statistic, error=function(x) NA ))
  return(t.obs)
}

#boot analysis with R=999 reps
start_time <- Sys.time()
boot.out <- boot(data = mat, statistic =boot.fun, R=999, parallel = "multicore", ncpus = 48)
Sys.time()-start_time
#> Time difference of 3.727077 mins


#vectorized format for matrix
mat.new <- data.table(mat %>% 
                           # Convert row names to a column
                           rownames_to_column('gene') %>% 
                           # "Gather" the data using the pivot_longer function (preferred to the old "gather" function)
                           pivot_longer(cols = -gene,
                                        names_to = 'cond',
                                        values_to = 'exp') %>% 
                           # Fix condition
                           mutate(cond = case_when(
                             str_detect(cond, pattern = 'C') ~ '1',
                             TRUE ~ '2')))
#> Error in rownames_to_column(., "gene"): could not find function "rownames_to_column"
head(mat.new, n=6)
#> Error in head(mat.new, n = 6): object 'mat.new' not found
mat.new <- mat.new[, list(data=list(data.table(cond, exp))), by=gene]
#> Error in eval(expr, envir, enclos): object 'mat.new' not found
head(mat.new, n=6)
#> Error in head(mat.new, n = 6): object 'mat.new' not found

#new function for vectorized bootstrap
bootFun_new <- function(df,f) {
  n <- nrow(df)
  xind <- which(df[, 1] == 2) #Treatment selection
  yind <- which(df[, 1] == 1) #Control selection
  nx <- length(xind)
  ny <- length(yind)
  new.df <- df
  new.df[xind, 2] <- df[xind, 2] - colMeans(df[xind, 2]) + colMeans(df[, 2])
  new.df[yind, 2] <- df[yind, 2] - colMeans(df[yind, 2]) + colMeans(df[, 2])
  df <- new.df
  MX <- sum(df[xind, 2] * f[xind])/sum(f[xind])
  SX <- sum(df[xind, 2]^2 * f[xind])/sum(f[xind]) - MX^2
  SX <- nx * SX/(nx - 1)
  MY <- sum(df[yind, 2] * f[yind])/sum(f[yind])
  SY <- sum(df[yind, 2]^2 * f[yind])/sum(f[yind]) - MY^2
  SY <- ny * SY/(ny - 1)
  SXY <- sqrt((SX/nx) + (SY/ny))
  (MX -MY)/SXY
}

#bootstrap for vectorized format with boot.fun_new
plan(multicore, workers=48)
mat.new[, booted := future_map(.x=data, ~ boot(data=.x, sim = "ordinary", 
                                                   statistic = bootFun_new, R = 999,
                                                   stype = "f", strata = .x$cond))]
```

<sup>Created on 2021-04-09 by the [reprex package](https://reprex.tidyverse.org) (v2.0.0)</sup>

**Comment by DavisVaughan** [comment]
I _think_ that the problem is either with the way you have rearranged your data, or with the new function you have created to use on the rearranged data.

I tried removing furrr altogether. In this tweaked example, I:

- Dropped the number of elements from 600000 to 18750
- Dropped R from 999 to 50
- Turned off parallel code everywhere

So now everything runs sequentially. In theory this should be approximately equal?

The `boot()` call runs in about 5 seconds, but the `map()` version runs for quite a while. I let it sit for a few minutes and it didn't finish.

So my guess is that it isn't furrr. Its just the fact that whatever you have changed to be able to call this in a `map()` style has made things significantly slower

```r
library(dplyr)
library(tidyr)
library(stringr)
library(purrr)
library(data.table)
library(boot)
library(tibble)

#Expression matrix
mat <- data.frame(matrix(runif(18750),ncol=30))
colnames(mat) <- c(paste0("C", seq(1:10)), paste0("T", seq(11:30)))
rownames(mat) <- c(paste0("Gene", seq_len(nrow(mat))))

#function for matrix format 
boot.fun <- function(data, i){
  data <- data[i,]
  t.obs <- apply(data, 1, function(x) tryCatch(t.test(x[11:30], x[1:10], alternative = "two.sided",
                                                      paired = FALSE, var.equal = FALSE)$statistic, error=function(x) NA ))
  return(t.obs)
}

#boot analysis with R=50 reps
start_time <- Sys.time()
boot.out <- boot(data = mat, statistic =boot.fun, R=50)
Sys.time()-start_time


#vectorized format for matrix
mat.new <- data.table(mat %>% 
                        # Convert row names to a column
                        rownames_to_column('gene') %>% 
                        pivot_longer(cols = -gene,
                                     names_to = 'cond',
                                     values_to = 'exp') %>% 
                        # Fix condition
                        mutate(cond = case_when(
                          str_detect(cond, pattern = 'C') ~ '1',
                          TRUE ~ '2')))

mat.new <- mat.new[, list(data=list(data.table(cond, exp))), by=gene]

#new function for vectorized bootstrap
bootFun_new <- function(df,f) {
  n <- nrow(df)
  xind <- which(df[, 1] == 2) #Treatment selection
  yind <- which(df[, 1] == 1) #Control selection
  nx <- length(xind)
  ny <- length(yind)
  new.df <- df
  new.df[xind, 2] <- df[xind, 2] - colMeans(df[xind, 2]) + colMeans(df[, 2])
  new.df[yind, 2] <- df[yind, 2] - colMeans(df[yind, 2]) + colMeans(df[, 2])
  df <- new.df
  MX <- sum(df[xind, 2] * f[xind])/sum(f[xind])
  SX <- sum(df[xind, 2]^2 * f[xind])/sum(f[xind]) - MX^2
  SX <- nx * SX/(nx - 1)
  MY <- sum(df[yind, 2] * f[yind])/sum(f[yind])
  SY <- sum(df[yind, 2]^2 * f[yind])/sum(f[yind]) - MY^2
  SY <- ny * SY/(ny - 1)
  SXY <- sqrt((SX/nx) + (SY/ny))
  (MX -MY)/SXY
}

start_time <- Sys.time()
mat.new[, booted := map(.x=data, ~ boot(data=.x, sim = "ordinary", 
                                               statistic = bootFun_new, R = 50,
                                               stype = "f", strata = .x$cond))]
Sys.time()-start_time
```

**Comment by Dokmen** [comment]
Thank you so much your response. The  example is in the boot pckage page. [https://github.com/cran/boot/blob/master/tests/Examples/boot-Ex.Rout.save](url) 
I changed only 1:8 lines in the "bootFun.new" function.  These lines belong adjusted algorith as described in  Efron and Tibshirani (1993), (Algorithm 16.2, p224)
The other lines was written by creater. This week I tried hard too and ran the command with Parlapply, removing lines 7-8-9 and splitting the vector with the "gene". The command run in a short time. However, since Boot works with functions, I think it is quite difficult to operate with map here, especially by boot stype and strata options.
